[IMP] sale_order_product_recommendation: Display only active or sale ok products in recommendation wizard

sergio-teruel opened the pull request:
cc @tecnativa TT19289 TT19978

rafaelbn reviewed (commented):
- Make sales to a partner to have products for the recomendation. Example SO123 with 3 products.
- Duplicate SO123, and the new is SO124, clic in Recommended product and you get:

Error:
Odoo Server Error

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 653, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 312, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 695, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 344, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 337, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 939, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 517, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/OCB-11.0/addons/web/controllers/main.py", line 934, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/odoo/OCB-11.0/addons/web/controllers/main.py", line 926, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/models.py", line 5222, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File "/.repo_requirements/odoo/odoo/models.py", line 5039, in _onchange_eval
    method_res = method(self)
  File "/home/odoo/build/OCA/sale-workflow/sale_order_product_recommendation/wizards/sale_order_recommendation.py", line 98, in _generate_recommendations
    found_line = found_dict[line.product_id.id]
KeyError: 31

yajo reviewed (approved):
I was wondering why this happened. TIL [`active_test` is ignored by `read()`](https://github.com/odoo/odoo/blob/cd1b56b41fb6e5b28d22e33867bea31343475fe8/odoo/models.py#L4270), and I guess it also happens with `read_group()`.

pedrobaeza commented:
Before merging, I would like a test in production of times before and after this patch.

yajo commented:
What possible bottlenecks do you see introduced here?

pedrobaeza commented:
An extra domain on submodel, that introduces by design of ORM a query with `IN` ids, which is slow with a lot of records.

sergio-teruel commented:
Sorry!!!! but I do not add any "IN" clause...

pedrobaeza commented:
I know, @sergio-teruel, it's not you, it's the ORM. Let me explain:

- You a have an ORM domain that is `[('product_id.active', '=', True)]`.
- Odoo queries `product_product` table for getting ids that are active. Imagine this returns `[1, 2, 4]`.
- Then Odoo performs this domain in the main table: `[('product_id', 'in', [1, 2, 4])]`. 
- That query is very slow with a high number of elements in the `in` part.

sergio-teruel commented:
Ohh!! Yeah... I understood!!

sergio-teruel commented:
@rafaelbn Fixed error done!!

pedrobaeza commented:
And why doing `filtered` afterwards? You should do one or the other, nor both.

sergio-teruel commented:
No, because if you have a sale order with products... after that you disable any product of this sale order... and click on recommendations the methos generated_recommendations try fill the wizard lines with the orderline products and others, so if any has been disabed you obtain key error...

pedrobaeza commented:
Well, I vote for removing then the domain part.

sergio-teruel commented:
No, but what do you think if I add this part of code in line https://github.com/OCA/sale-workflow/pull/930/files#diff-bf4bc078958d60f65a62f59551d2e6e1R114

`product = self.env['product.product'].browse(line["product_id"][0])`
`if not product.active or not product.sale_ok:`
`continue`

sergio-teruel commented:
I don't see disable products in recommended lines

pedrobaeza commented:
OK, as I'm on holidays, I let the rest of the reviewers to decide the final patch. I just wanted to warn about the performance impact.

yajo commented:
I'm not sure about the underlying query result, but it would be strange for me that it does that `IN` stuff. That query is quite easy for SQL and can be solved with a `JOIN` and/or a `WHERE` clause. Maybe you can enable sql query logging and see the result. If it were to use a `IN`, theoretically that wouldn't have performance impact in a postgres server that has `work_mem = 32MB`, so again probably not harming.

*Just thinking out loud, I have not analyzed this in deep.

yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -87,8 +91,9 @@ def _generate_recommendations(self):
-        for line in self.order_id.order_line:
+        # Add products from sale order lines that are active and can be sold
+        for line in self.order_id.order_line.filtered(
+                lambda x: x.product_id.active and x.product_id.sale_ok):
```
```suggestion
                lambda x: x.product_id.active and x.product_id.sale_ok):
            if not (line.product_id.active and line.product_id.sale_ok):
                continue
```

This should avoid double-looping.

sergio-teruel replied:
In this loop I have actives and sale_or products, so I do nor need this condition

yajo replied:
Yes, I mean that you can remove the `.filtered()` and use a normal condition instead, to avoid double-looping.

Other wise Odoo loops once for the filtered and a 2nd time for the rest of execution.

pedrobaeza commented:
`JOIN` is only done if the relation field has `auto_join=True`. The explanation is that if not, record rules are not applied (as well as active field ORM filter).

yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -100,9 +105,13 @@ def _generate_recommendations(self):
         # Add those recommendations too
+        products_allowed_ids = self.env['product.product'].browse(
+            [l["product_id"][0] for l in found_lines]).filtered(
+            lambda x: x.active and x.sale_ok).ids
```
I'd say this makes more sense here:

```python
products_allowed_ids = set(self.env["product.product"].search([
    ("id", "in", [l["product_id"][0] for l in found_lines]),
    ("id", "not in", existing_product_ids),
    ("active", "=", True),
    ("sale_ok", "=", True),
).ids)
```

You can even remove the `active=true` test because it's automatic when you're filtering on the main model. I left it just to let you understand better the change.

If you're gonna make a query, then at least you save the latter python-side filtering.

Yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -100,9 +105,13 @@ def _generate_recommendations(self):
         i = 0
         for line in found_lines:
-            if line["product_id"][0] in existing_product_ids:
+            if (line["product_id"][0] in existing_product_ids or
```
If you do the above, this `if` should change accordingly.

Yajo replied:
```suggestion
            if line["product_id"][0] not in products_allowed_ids:
```

You don't need now the existing_product_ids check because it's applied already in the domain above.

yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -46,6 +46,14 @@ class SaleOrderRecommendation(models.TransientModel):
+            '|',
+            ("qty_delivered", "!=", 0.0),
+            ("order_id", "=", self.order_id.id),
+            ]
```
```suggestion
        ]
```

sergio-teruel commented:
Changes done

rafaelbn reviewed (approved):
:+1:

yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -46,6 +46,14 @@ class SaleOrderRecommendation(models.TransientModel):
     def _default_order_id(self):
         return self.env.context.get("active_id", False)
 
+    def _prepare_sale_order_line_domain(self, sales):
```
I think this title is a bit misleading.

```suggestion
    def _prepare_sale_order_line_recommendations_domain(self, sales):
```

yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -100,9 +106,15 @@ def _generate_recommendations(self):
+        products_allowed_ids = set(self.env["product.product"].search([
+            ("id", "in", [l["product_id"][0] for l in found_lines]),
+            ("id", "not in", existing_product_ids),
+            ("sale_ok", "=", True),
```
I think you're repeating this part, which is done at line 96.

```suggestion
```

yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -100,9 +106,15 @@ def _generate_recommendations(self):
         for line in found_lines:
-            if line["product_id"][0] in existing_product_ids:
+            if (line["product_id"][0] in existing_product_ids or
+                    line["product_id"][0] not in products_allowed_ids):
```
This should be:

```suggestion
            if line["product_id"][0] not in products_allowed_ids:
```

yajo commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -67,11 +75,7 @@ def _generate_recommendations(self):
-                '|', ("qty_delivered", "!=", 0.0),
-                     ("order_id", "=", self.order_id.id),
-            ],
+            self._prepare_sale_order_line_domain(sales),
```
```suggestion
            self._prepare_sale_order_line_recommendations_domain(sales),
```

sergio-teruel commented:
@Yajo please, Can you takes up this job?? thanks

sergio-teruel commented:
TT19289

yajo commented:
/ocabot merge patch

pedrobaeza commented:
Wait a moment! I have deleted the merging branch.

This is installed in several customers and we are modifying the behavior, which can be not desirable for the rest. We should introduce a flag for allowing to preserve previous behavior.

yajo commented:
It seems that all our customers agree that this is the expected behavior, so we can safely merge.

pedrobaeza commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -46,6 +46,15 @@ class SaleOrderRecommendation(models.TransientModel):
     def _default_order_id(self):
         return self.env.context.get("active_id", False)
 
+    def _recomendable_sale_order_lines_domain(self, so_ids):
```
Please pass always that you are able to recordsets, not ids

pedrobaeza commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -46,6 +46,15 @@ class SaleOrderRecommendation(models.TransientModel):
+        return [
+            ("order_id", "!=", self.order_id.id),
+            ("order_id", "in", so_ids),
+            ("product_id.active", "=", True),
```
This is redundant, as Odoo already do that

pedrobaeza commented:
Now Travis is red

yajo commented:
:man_facepalming: I removed one thing with dates thinking I was in v12, sorry, it's back.

I have also removed these lines now: https://github.com/OCA/sale-workflow/blob/cc0a9f2e06480bf8c564b748ee37512dba958866/sale_order_product_recommendation/wizards/sale_order_recommendation.py#L96-L97

I think it's a mistake. I didn't understand what was this doing, but now that I do, it shouldn't be there.

This process was doing:

1. Find all products already in the current SO.
1. Filter out those that are not `active` or `sale_ok`.
1. Put them all on top of the recommendation list.
1. Find recent products in recent SO.
1. Add them as recommendations.

Step 2 seems wrong IMHO because, if for any chance the product that shouldn't be saleable is found in the present SO, you still want it there to be able to use the add/remove buttons easily.

What do you think @sergio-teruel?

pedrobaeza commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -46,6 +46,23 @@ class SaleOrderRecommendation(models.TransientModel):
+            ("date_order", ">=", start),
+        ])
+        return [
+            ("order_id", "!=", self.order_id.id),
```
Don't do this for performance, but be sure `self.order_id` is not in `other_sales`

yajo replied:
I don't understand... What's the performance problem of excluding an ID? That's an indexed column and should be blazing fast... :confused:

pedrobaeza replied:
You are using 2 times `order_id`. Just do it once.

pedrobaeza commented on sale_order_product_recommendation/wizards/sale_order_recommendation.py:
```diff
@@ -46,6 +46,23 @@ class SaleOrderRecommendation(models.TransientModel):
+        return [
+            ("order_id", "!=", self.order_id.id),
+            ("order_id", "in", other_sales.ids),
+            ("product_id.active", "=", True),
```
As most of the products will be active and probably `sale_ok`, I would bet for doing this in a later filter, but I'm not sure if you have already tried this and measure performance

yajo replied:
Given we have to ask the DB for a recordset, it's better to filter it directly in the DB. It's always faster than downloading extra recordsets just to ignore them later in Python code.

If the recordsets were already cached, that could be a different story, but it's not the case here.

pedrobaeza replied:
The problem is that this is not auto-joined, so it actually requires several queries + the problem with `in` operator in SQL a list with thousand of elements.

yajo replied:
I have been profiling this.

I didn't understand what you were saying, but now I see that indeed these filters are made in separate SQL queries. That's quite strange, since it could just be added to the same query. I guess that's to apply permissions...

Anyways, removing these leafs from the domain we save 2 queries and ~7ms in total, but later we'd have to browse all product records and manually prefetch their `active` and `sale_ok` fields to actually end up saving 1 query, at the cost of code readability.

IMHO it's not worth it.

pedrobaeza replied:
We would save that queries if `product_id` has `auto_join=True`, but that disables ACL check. In this case, maybe it makes sense to add it. In current state, the problem is not the extra 2 queries, but having thousand of product, creating the third query: `('product_id', 'in', [<thousand of ids>])`, which the slowest one. Have you tried?

yajo replied:
I haven't tried that, but I doubt that is a performance problem.

I think you are in a confusion with the performance problems we have faced in the past with subqueries, not for lists of ids. I mean, it's more things like `where id in (select id from lalala)`, because if the server has not enough worker memory, it executes the subquery record by record and the execution time grows exponentially. But that's a completely different case.

Even in Python, if you use the `in` operator with a `set()`, you get instant results no matter the set size. I imagine Postgres uses a similar optimization under the hood to avoid that problem.

I'd say the best we can do is merge and then patch it if we really see some real bottleneck in production. Otherwise we might be wasting time in fixing something that is not broken.

pedrobaeza replied:
Seen directly in a meeting and it doesn't seem to affect the performance, so we keep it this way.

yajo committed:
[IMP] sale_order_product_recommendation: Display only active or sale ok products in recommendation wizard

BTW I fixed a bug that counted products in the active SO as delivered, when it was impossible in most cases. Fixed the test to reflect the new logic.

Co-Authored-By: Jairo Llopis <Yajo@users.noreply.github.com>

pedrobaeza commented:
/ocabot merge minor

pedrobaeza commented:
Cherry-picked and adapted for v12 in 40f8e9e39e000292b4b44e43d7d01ae4c115c2a1 and v13 in 0614cb6f7975c14563d431e983e91c1e984b27af
